src/drt/src/dr/FlexDR.cpp

```diff
@@ -104,6 +173,151 @@ int FlexDRWorker::main(frDesign* design)
   return 0;
 }
 
+inline frBlockObject* getEqObject(frBlockObject* srObj, frDesign* design)
+{
+  switch (srObj->typeId()) {
+    case frcNet: {
+      frNet* srNet = (static_cast<frNet*>(srObj));
+      auto net = design->getTopBlock()->findNet(srNet->getName());
+      return net;
+    }
+    case frcInstTerm: {
+      frInstTerm* srTerm = (static_cast<frInstTerm*>(srObj));
+      auto srInst = srTerm->getInst();
+      auto inst = design->getTopBlock()->findInst(srInst->getName());
+      if (inst == nullptr)
+        return nullptr;
+      for (auto& term : inst->getInstTerms())
+        if (term->getId() == srTerm->getId())
+          return term.get();
+      return nullptr;
+    }
+    case frcTerm: {
+      frTerm* srTerm = (static_cast<frTerm*>(srObj));
+      return design->getTopBlock()->getTerm(srTerm->getName());
+    }
+    case frcInstBlockage: {
+      frInstBlockage* srBlkg = (static_cast<frInstBlockage*>(srObj));
+      auto srInst = srBlkg->getInst();
+      auto inst = design->getTopBlock()->findInst(srInst->getName());
+      if (inst == nullptr)
+        return nullptr;
+      for (auto& blkg : inst->getInstBlockages())
+        if (blkg->getId() == srBlkg->getId())
+          return blkg.get();
+      return nullptr;
+    }
+    case frcBlockage: {
+      frBlockage* srBlkg = (static_cast<frBlockage*>(srObj));
+      for (auto& blkg : design->getTopBlock()->getBlockages())
+        if (blkg->getId() == srBlkg->getId())
+          return blkg.get();
+    }
+    default:
+      return nullptr;
```

Comment:
Seems like we should change the return type to std::optional so that downstream clients are aware this can return null. 

https://en.cppreference.com/w/cpp/utility/optional

Reply:
I'd like to use this, but it's never used in tritonroute. Many functions have the same behavior of returning a nullptr. I am not sure if it's gonna look alien in the code, or it would be a starting point for others to use it in the future. Also I think this should be used more in public APIs with multiple users rather than a simple inline function with limited users.

Reply:
I'll leave thr final call to you, but I think optional still provides value because it enforces a null check on consumers which makes the code more resistant and provides a chaining function that can improve readability.

I think optional is one of those things that can be added incrementally while not having to effect other code. We can gradually transition TritonRoute to optional if desired

Reply:
I think this is worth taking up in another PR but this one is already overdue.